accessors/spanner/spanner_accessor.go

```diff
@@ -0,0 +1,203 @@
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	"cloud.google.com/go/spanner/admin/database/apiv1/databasepb"
+	"cloud.google.com/go/spanner/admin/instance/apiv1/instancepb"
+	spanneradmin "github.com/GoogleCloudPlatform/spanner-migration-tool/accessors/clients/spanner/admin"
+	spannerclient "github.com/GoogleCloudPlatform/spanner-migration-tool/accessors/clients/spanner/client"
+	spinstanceadmin "github.com/GoogleCloudPlatform/spanner-migration-tool/accessors/clients/spanner/instanceadmin"
+	"github.com/GoogleCloudPlatform/spanner-migration-tool/common/utils"
+	"google.golang.org/api/iterator"
+)
+
+// The SpannerAccessor provides methods that internally use a spanner client (can be adminClient/databaseclient/instanceclient etc).
+// Methods should only contain generic logic here that can be used by multiple workflows.
+type SpannerAccessor interface {
+	// Fetch the dialect of the spanner database.
+	GetDatabaseDialect(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string) (string, error)
+	// CheckExistingDb checks whether the database with dbURI exists or not.
+	// If API call doesn't respond then user is informed after every 5 minutes on command line.
+	CheckExistingDb(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string) (bool, error)
+	// Create a database with no schema.
+	CreateEmptyDatabase(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string) error
+	// Fetch the leader of the Spanner instance.
+	GetSpannerLeaderLocation(ctx context.Context, instanceClient spinstanceadmin.InstanceAdminClient, instanceURI string) (string, error)
+	// Check if a change stream already exists.
+	CheckIfChangeStreamExists(ctx context.Context, changeStreamName, dbURI string) (bool, error)
+	// Validate that change stream option 'VALUE_CAPTURE_TYPE' is 'NEW_ROW'.
+	ValidateChangeStreamOptions(ctx context.Context, changeStreamName, dbURI string) error
+	// Create a change stream with default options.
+	CreateChangeStream(ctx context.Context, adminClient spanneradmin.AdminClient, changeStreamName, dbURI string) error
+}
+
+// This implements the SpannerAccessor interface. This is the primary implementation that should be used in all places other than tests.
+type SpannerAccessorImpl struct{}
+
+func (sp *SpannerAccessorImpl) GetDatabaseDialect(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string) (string, error) {
+	result, err := adminClient.GetDatabase(ctx, &databasepb.GetDatabaseRequest{Name: dbURI})
+	if err != nil {
+		return "", fmt.Errorf("cannot connect to database: %v", err)
+	}
+	return strings.ToLower(result.DatabaseDialect.String()), nil
+}
+
+func (sp *SpannerAccessorImpl) CheckExistingDb(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string) (bool, error) {
+	gotResponse := make(chan bool)
+	var err error
+	go func() {
+		_, err = adminClient.GetDatabase(ctx, &databasepb.GetDatabaseRequest{Name: dbURI})
```

Comment:
why do we need to launch a goroutine for this? why not just normally fetchif the database exists?

Reply:
I think this was done via a goroutine to print a warning after 5 mins of no response on ln73.
I have just lifted the function as is from conversion.go.